crates/ursa-consensus/src/consensus.rs

```diff
@@ -0,0 +1,268 @@
+
+        // Start the timer to signal when your node thinks its ready to change epochs
+        let time_until_epoch_change = Duration::from_millis(until_epoch_ends);
+        self.wait_to_signal_epoch_change(time_until_epoch_change)
+            .await;
+
+        service.start(self.execution_state.clone()).await;
+
+        *self.epoch_state.lock().await = Some(EpochState { narwhal: service })
+    }
+
+    async fn move_to_next_epoch(&self) {
+        {
+            let epoch_state_mut = self.epoch_state.lock().await.take();
+            if let Some(state) = epoch_state_mut {
+                state.narwhal.shutdown().await
+            }
+        }
+        self.start_current_epoch().await
+    }
+
+    async fn wait_to_signal_epoch_change(&self, time_until_change: Duration) {
+        let primary_public_key = self.narwhal_args.primary_keypair.public().clone();
+        let mempool_address = self.mempool_address.clone();
+        task::spawn(async move {
+            time::sleep(time_until_change).await;
+            // We shouldnt panic here lets repeatedly try
+            loop {
+                time::sleep(Duration::from_secs(1)).await;
+
+                let txn = match serde_json::to_vec(&encode_signal_epoch_call(
+                    primary_public_key.to_string(),
+                )) {
+                    Ok(txn) => txn,
+                    Err(_) => {
+                        error!("Error signaling epoch change, trying again");
+                        continue;
+                    }
+                };
+
+                let request = TransactionProto {
+                    transaction: Bytes::from(txn),
+                };
+
+                let mut client = match TransactionsClient::connect(mempool_address.clone()).await {
+                    Ok(client) => client,
+                    Err(e) => {
+                        error!("Error building client to signal epoch change {:?}", e);
+                        continue;
+                    }
+                };
+
+                if client.submit_transaction(request).await.is_ok() {
+                    break;
+                }
+                error!("Error signaling epoch change trying again");
+            }
+        });
+    }
+
+    pub async fn start(&mut self) {
+        self.start_current_epoch().await;
+        loop {
+            let reconfigure_future = self.reconfigure_notify.notified();
+            let shutdown_future = self.shutdown_notify.notified();
+            pin!(shutdown_future);
+            pin!(reconfigure_future);
+            select! {
+                _ = shutdown_future => {
+                    break
+                }
+                _ = reconfigure_future => {
+                    self.move_to_next_epoch().await;
+                    continue
+                }
+            }
+        }
+    }
+
+    pub async fn shutdown(&mut self) {
+        self.shutdown_notify.notify_waiters();
+    }
+
+    pub async fn run_edge_node(&self) {
```

Comment:
The ursa node would be run from here?

Reply:
Maybe its bad naming and its pretty much a placeholder for now. but this is the entry to non validator logic that wouldnt be running narwhal.